### ISSUE TITLE
GELF forwarding in rsyslog: rsyslog is unable to use a null byte as a message delimiter

### ISSUE THREAD

**Issue by deoren** [rsyslog/rsyslog-doc#621]
From the [GELF forwarding in rsyslog](https://github.com/rsyslog/rsyslog-doc/blob/c0fb71c34ac57f88a872c26f902d54661b2621ab/source/tutorials/gelf_forwarding.rst) doc:

> We now have a syslog forwarding action. This uses the omfwd module. Please note that the case above only works for UDP transport. **When using TCP, Graylog expects a Nullbyte as message delimiter. This is currently not possible with rsyslog.**

Is this still true?

**Comment by rgerhards** [comment]
> Is this still true?

I think so, but if important we should probably test or review code (unfortunately I have no time at the moment to do this).

**Comment by deoren** [comment]
Understood. I should have probably phrased my question to be a little more focused:

*Is rsyslog capable of using a null byte as a delimiter?*

If you managed to catch the direction of my too short question and that is what you were responding to, not a problem. I can always check with the Graylog community if digging further into their docs doesn't yield a direct answer. It may be that an older version had that requirement whereas the newer does not.

Regarding Graylog receiving over TCP, we're using omfwd with this template and are getting good results:

```
# https://github.com/Graylog2/graylog-guide-syslog-linux
template(name="graylogrfc5424" type="list") {
    constant(value="<")
    property(name="pri")
    constant(value=">")
    property(name="protocol-version")
    constant(value=" ")
    property(name="timestamp" dateFormat="rfc3339")
    constant(value=" ")
    property(name="hostname")
    constant(value=" ")
    property(name="app-name")
    constant(value=" ")
    property(name="procid")
    constant(value=" ")
    property(name="msgid")
    constant(value=" ")
    property(name="structured-data")
    constant(value=" ")
    property(name="msg" droplastlf="on" )
    constant(value="\n")
}
```

I hope to make time later this year to start submitting log messages to Graylog via JSON. Seeing the recent improvements to JSON-based message handling in rsyslog is very encouraging in that regard. :)

**Comment by rgerhards** [comment]
The NUL terminator cannot be used inside a template. However, we could still permit to add it as a delimiter in code. But I think we have not yet done this. 99% sure.

**Comment by dch** [comment]
what about enabling a `\u00` or `\0` in a constant field? e.g.:

    constant(value="\0")

But is it even possible to append that to the end of an existing json template? One would assume not.

**Comment by davidelang** [comment]
On Thu, 22 Nov 2018, Dave Cottlehuber wrote:

> what about enabling a `\u00` or `\0` in a constant field? e.g.:
>
>    constant(value="\0")
>
> But is it even possible to append that to the end of an existing json template? One would assume not.

Rsyslog uses c strings internally, in them a null indicates the end of the 
string, so I don't think that rsyslog is able to send null delimited messages.

David Lang

**Comment by elcamlost** [comment]
What about [TCP_FrameDelimiter](https://github.com/rsyslog/rsyslog-doc/blob/master/source/configuration/modules/omfwd.rst#tcp_framedelimiter) option in omfwd?
I configured omwd as 
```
ruleset(name="graylog-ext") {
  action(
    type="omfwd"
    Target="graylog"
    Port="12201"
    Protocol="tcp"
    KeepAlive="on"
    template="gelf"
    StreamDriver="gtls"
    StreamDriverMode="1"
    StreamDriverAuthMode="x509/name"
    StreamDriverPermittedPeers="graylog"
    TCP_FrameDelimiter="0"
  )
}
```

And everything seems to work fine.

**Comment by mchubby** [comment]
Hello,
Please close that issue, this was addressed by #840 in docs
and implemented in v8.28 back in 2017 -- [1610](https://github.com/rsyslog/rsyslog/pull/1610) ([1602](https://github.com/rsyslog/rsyslog/issues/1602)).

**Comment by rgerhards** [comment]
thx for the heads-up
